File: backend/app/core/secrets.py
import os
import secrets
import logging
from dotenv import load_dotenv

# Load environment variables from .env file if it exists
load_dotenv()

logger = logging.getLogger(__name__)

REQUIRED_ENV_VARS = {
    "DATABASE_URL":          {"min_length": 20},
    "REDIS_URL":             {"min_length": 10},
    "JWT_SECRET":            {"min_length": 64},
    "RAZORPAY_KEY_ID":       {"min_length": 14},
    "RAZORPAY_KEY_SECRET":   {"min_length": 20},
    "RAZORPAY_WEBHOOK_SECRET": {"min_length": 20},
    "DB_ENCRYPTION_KEY":     {"min_length": 32},
    "ENVIRONMENT":           {"allowed": ["production",
                                          "staging",
                                          "development"]},
}

def validate_all_secrets():
    errors = []
    env = os.getenv("ENVIRONMENT", "development")

    # For development/staging, we can auto-fill missing/short variables with secure/valid dummy fallbacks
    if env != "production":
        if not os.getenv("JWT_SECRET") or len(os.getenv("JWT_SECRET")) < 64:
            os.environ["JWT_SECRET"] = secrets.token_hex(32)  # Generates 64 chars hex string
            logger.warning("Generated fallback JWT_SECRET for non-production environment.")

        if not os.getenv("DB_ENCRYPTION_KEY") or len(os.getenv("DB_ENCRYPTION_KEY")) < 32:
            os.environ["DB_ENCRYPTION_KEY"] = secrets.token_hex(16)  # Generates 32 chars hex string
            logger.warning("Generated fallback DB_ENCRYPTION_KEY for non-production environment.")

        if not os.getenv("RAZORPAY_KEY_ID") or len(os.getenv("RAZORPAY_KEY_ID")) < 14:
            os.environ["RAZORPAY_KEY_ID"] = "rzp_test_" + secrets.token_hex(5)
            logger.warning("Generated fallback RAZORPAY_KEY_ID for non-production environment.")

        if not os.getenv("RAZORPAY_KEY_SECRET") or len(os.getenv("RAZORPAY_KEY_SECRET")) < 20:
            os.environ["RAZORPAY_KEY_SECRET"] = "rzp_test_secret_" + secrets.token_hex(4)
            logger.warning("Generated fallback RAZORPAY_KEY_SECRET for non-production environment.")

        if not os.getenv("RAZORPAY_WEBHOOK_SECRET") or len(os.getenv("RAZORPAY_WEBHOOK_SECRET")) < 20:
            os.environ["RAZORPAY_WEBHOOK_SECRET"] = "rzp_webhook_secret_" + secrets.token_hex(4)
            logger.warning("Generated fallback RAZORPAY_WEBHOOK_SECRET for non-production environment.")

        if not os.getenv("DATABASE_URL") or len(os.getenv("DATABASE_URL")) < 20:
            # Fallback to local SQLite URL
            os.environ["DATABASE_URL"] = "sqlite:///./sql_app.db"
            logger.warning("Fallback to local SQLite database in development.")

        if not os.getenv("REDIS_URL") or len(os.getenv("REDIS_URL")) < 10:
            os.environ["REDIS_URL"] = "redis://localhost:6379/0"
            logger.warning("Fallback to localhost Redis in development.")

    for var, rules in REQUIRED_ENV_VARS.items():
        value = os.getenv(var, "")

        if not value:
            errors.append(f"Missing: {var}")
            continue

        min_len = rules.get("min_length", 0)
        if len(value) < min_len:
            errors.append(
                f"{var} too short: {len(value)} < {min_len}"
            )

        allowed = rules.get("allowed")
        if allowed and value not in allowed:
            errors.append(
                f"{var} invalid value: {value}"
            )

    if errors:
        raise ValueError(
            "Secret validation failed:\n" +
            "\n".join(errors)
        )

    logger.info("All secrets validated successfully")

File: streamguard-ai/backend/src/services/secrets.js
import crypto from 'crypto';
import winston from 'winston';

const logger = winston.createLogger({
  transports: [new winston.transports.Console()]
});

const REQUIRED_ENV_VARS = {
  "DATABASE_URL": { minLength: 20 },
  "REDIS_URL": { minLength: 10 },
  "SUPABASE_URL": { minLength: 20 },
  "SUPABASE_ANON_KEY": { minLength: 30 },
  "DB_ENCRYPTION_KEY": { minLength: 32 },
  "RAZORPAY_KEY_ID": { minLength: 14 },
  "RAZORPAY_KEY_SECRET": { minLength: 20 },
  "RAZORPAY_WEBHOOK_SECRET": { minLength: 20 },
  "ENVIRONMENT": { allowed: ["production", "staging", "development"] },
};

export function validateAllSecrets() {
  const env = process.env.ENVIRONMENT || "development";
  const errors = [];

  // Generate safe development fallbacks
  if (env !== "production") {

    if (!process.env.DB_ENCRYPTION_KEY || process.env.DB_ENCRYPTION_KEY.length < 32) {
      process.env.DB_ENCRYPTION_KEY = crypto.randomBytes(16).toString('hex'); // 32 chars
      logger.warn("Generated fallback DB_ENCRYPTION_KEY for development.");
    }
    if (!process.env.RAZORPAY_KEY_ID || process.env.RAZORPAY_KEY_ID.length < 14) {
      process.env.RAZORPAY_KEY_ID = "rzp_test_" + crypto.randomBytes(5).toString('hex');
      logger.warn("Generated fallback RAZORPAY_KEY_ID for development.");
    }
    if (!process.env.RAZORPAY_KEY_SECRET || process.env.RAZORPAY_KEY_SECRET.length < 20) {
      process.env.RAZORPAY_KEY_SECRET = "rzp_test_secret_" + crypto.randomBytes(4).toString('hex');
      logger.warn("Generated fallback RAZORPAY_KEY_SECRET for development.");
    }
    if (!process.env.RAZORPAY_WEBHOOK_SECRET || process.env.RAZORPAY_WEBHOOK_SECRET.length < 20) {
      process.env.RAZORPAY_WEBHOOK_SECRET = "rzp_webhook_secret_" + crypto.randomBytes(4).toString('hex');
      logger.warn("Generated fallback RAZORPAY_WEBHOOK_SECRET for development.");
    }
    if (!process.env.DATABASE_URL || process.env.DATABASE_URL.length < 20) {
      process.env.DATABASE_URL = "postgresql://localhost:5432/flowshield";
      logger.warn("Fallback database URL configured to localhost PostgreSQL.");
    }
    if (!process.env.REDIS_URL || process.env.REDIS_URL.length < 10) {
      process.env.REDIS_URL = "redis://localhost:6379/0";
      logger.warn("Fallback Redis URL configured to localhost.");
    }
    if (!process.env.SUPABASE_URL || process.env.SUPABASE_URL.length < 20) {
      process.env.SUPABASE_URL = "https://mockproject.supabase.co";
      logger.warn("Fallback Supabase URL configured.");
    }
    if (!process.env.SUPABASE_ANON_KEY || process.env.SUPABASE_ANON_KEY.length < 50) {
      process.env.SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im1vY2twcm9qZWN0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE2Nzg4MDk2MDAsImV4cCI6MTk5NDM4NTYwMH0." + crypto.randomBytes(32).toString('base64');
      logger.warn("Fallback Supabase Anon Key configured.");
    }
  }

  // Perform checks
  for (const [varName, rules] of Object.entries(REQUIRED_ENV_VARS)) {
    const value = process.env[varName] || "";

    if (!value) {
      errors.append ? errors.push(`Missing: ${varName}`) : errors.push(`Missing: ${varName}`);
      continue;
    }

    if (rules.minLength && value.length < rules.minLength) {
      errors.push(`${varName} too short: ${value.length} < ${rules.minLength}`);
    }

    if (rules.allowed && !rules.allowed.includes(value)) {
      errors.push(`${varName} has invalid value: ${value}`);
    }
  }

  if (errors.length > 0) {
    throw new Error("Secret validation failed:\n" + errors.join("\n"));
  }

  logger.info("All environment secrets validated successfully.");
}
